File: android/app/src/main/java/com/eastlakestudio/stillhere/data/CareStore.kt
package com.eastlakestudio.stillhere.data

import android.content.Context
import android.content.SharedPreferences
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.temporal.ChronoUnit
import java.util.Calendar
import java.util.UUID
import java.util.concurrent.TimeUnit

/**
 * 一条关心关系
 *
 * 对应 iOS CareRelation
 */
data class CareRelation(
    val id: String = UUID.randomUUID().toString(),
    val name: String,          // 昵称
    val bindCode: String,      // 关心码（6位）
    val bindDate: Long = System.currentTimeMillis(),
    val lastActive: Long? = null,       // 对方最近活跃时间（Unix 秒，来自服务端）
    val isCharging: Boolean = false     // 对方是否在充电
) {
    /** 关心天数（从绑定日算起，最少 1 天，仅比较日期忽略时分秒） */
    val days: Int
        get() {
            val bindLocalDate = Instant.ofEpochMilli(bindDate).atZone(ZoneId.systemDefault()).toLocalDate()
            val today = LocalDate.now()
            val diff = ChronoUnit.DAYS.between(bindLocalDate, today).toInt()
            return diff.coerceAtLeast(0) + 1
        }

    /** 活动状态文案 */
    val activityText: String
        get() {
            val ts = lastActive ?: return "暂无活动"
            val secondsAgo = (System.currentTimeMillis() / 1000) - ts
            return when {
                secondsAgo < 60 -> "刚刚活跃"
                secondsAgo < 3600 -> "${secondsAgo / 60} 分钟前活跃"
                secondsAgo < 86400 -> "${secondsAgo / 3600} 小时前活跃"
                else -> "${secondsAgo / 86400} 天前活跃"
            }
        }

    /** 上次活动相对时间（不含"活跃"后缀，用于卡片第三行） */
    val lastActiveText: String
        get() {
            val ts = lastActive ?: return "暂无活动记录"
            val secondsAgo = (System.currentTimeMillis() / 1000) - ts
            return when {
                secondsAgo < 60 -> "刚刚"
                secondsAgo < 3600 -> "${secondsAgo / 60} 分钟前"
                secondsAgo < 86400 -> "${secondsAgo / 3600} 小时前"
                else -> "${secondsAgo / 86400} 天前"
            }
        }

    val isActive: Boolean
        get() {
            val ts = lastActive ?: return false
            return (System.currentTimeMillis() / 1000) - ts < 86400 // 24小时内算活跃
        }
}

/**
 * 关心关系本地持久化管理
 *
 * 对应 iOS CareStore ObservableObject
 */
class CareStore(context: Context) {

    private val prefs: SharedPreferences =
        context.getSharedPreferences("anhao.care", Context.MODE_PRIVATE)
    private val gson = Gson()
    private val scope = CoroutineScope(Dispatchers.Main)

    /** 被关心人数（来自服务端心跳响应，每次上报更新） */
    val caredByCountFlow: StateFlow<Int> = Reporter.caredByCount

    /** 我关心的人（关心列表） */
    private val _caring = MutableStateFlow<List<CareRelation>>(emptyList())
    val caring: StateFlow<List<CareRelation>> = _caring.asStateFlow()

    init {
        load()
        syncFromServer()
    }

    // MARK: - 从服务端恢复

    /** 新装 App 或清理数据后，从服务端拉回关心关系 */
    private fun syncFromServer() {
        scope.launch(Dispatchers.IO) {
            try {
                val remote = Reporter.fetchCaring()
                if (remote.isEmpty()) return@launch
                val existingCodes = _caring.value.map { it.bindCode }.toSet()
                val newOnes = remote.filter { it.bindCode !in existingCodes }
                if (newOnes.isEmpty()) return@launch
                _caring.value = _caring.value + newOnes.map {
                    CareRelation(name = it.bindCode, bindCode = it.bindCode)
                }
                save()
                android.util.Log.d("CareStore", "synced ${newOnes.size} relations from server")
            } catch (e: Exception) {
                android.util.Log.e("CareStore", "syncFromServer failed: ${e.message}")
            }
        }
    }

    // MARK: - 我关心

    fun addCaring(name: String, bindCode: String) {
        _caring.value = _caring.value + CareRelation(name = name, bindCode = bindCode)
        save()
        scope.launch {
            Reporter.registerCare(bindCode)
        }
        com.eastlakestudio.stillhere.StillHereApp.instance.syncToCloud()
    }

    fun removeCaring(relation: CareRelation) {
        _caring.value = _caring.value.filter { it.id != relation.id }
        save()
        scope.launch {
            Reporter.unregisterCare(relation.bindCode)
        }
        com.eastlakestudio.stillhere.StillHereApp.instance.syncToCloud()
    }

    fun updateCaringName(relation: CareRelation, newName: String) {
        val trimmed = newName.trim()
        if (trimmed.isEmpty()) return
        _caring.value = _caring.value.map {
            if (it.id == relation.id) it.copy(name = trimmed) else it
        }
        save()
        com.eastlakestudio.stillhere.StillHereApp.instance.syncToCloud()
    }

    fun updateCaringNameByCode(bindCode: String, name: String) {
        val existing = _caring.value.find { it.bindCode == bindCode }
        if (existing != null) {
            _caring.value = _caring.value.map {
                if (it.bindCode == bindCode) it.copy(name = name) else it
            }
            save()
        }
    }

    // MARK: - 活动状态刷新

    /** 拉取被关心者的最近活动状态 */
    fun refreshCaredStatus() {
        val codes = _caring.value.map { it.bindCode }.distinct()
        if (codes.isEmpty()) return
        scope.launch {
            try {
                val statuses = Reporter.fetchCaredStatus(codes)
                _caring.value = _caring.value.map { rel ->
                    val s = statuses[rel.bindCode]
                    if (s != null) {
                        rel.copy(lastActive = s.lastActive, isCharging = s.isCharging)
                    } else {
                        rel
                    }
                }
                save()
            } catch (e: Exception) {
                android.util.Log.e("CareStore", "refreshCaredStatus failed: ${e.message}")
            }
        }
    }

    // MARK: - 聚合

    val caringCount: Int get() = _caring.value.size

    /** 被关心天数（基于服务端返回的被关心人数计算起始日） */
    fun totalCaredDays(count: Int): Int {
        if (count <= 0) return 0
        val earliest = _caring.value.minOfOrNull { it.bindDate } ?: System.currentTimeMillis()
        return daysSince(earliest)
    }

    val totalCaringDays: Int
        get() {
            val earliest = _caring.value.minOfOrNull { it.bindDate } ?: return 0
            return daysSince(earliest)
        }

    private fun daysSince(timestamp: Long): Int {
        val bindCal = Calendar.getInstance().apply { timeInMillis = timestamp }
        val nowCal = Calendar.getInstance()
        var days = 0
        while (bindCal.before(nowCal)) {
            bindCal.add(Calendar.DAY_OF_MONTH, 1)
            days++
        }
        return maxOf(days, 0) + 1
    }

    // MARK: - 持久化

    private fun save() {
        prefs.edit()
            .putString("caring", gson.toJson(_caring.value))
            .apply()
    }

    private fun load() {
        val type = object : TypeToken<List<CareRelation>>() {}.type
        try {
            prefs.getString("caring", null)?.let {
                _caring.value = gson.fromJson(it, type)
            }
        } catch (_: Exception) { }
    }
}

File: ios/AnhaoSpike/AnhaoSpike/CareStore.swift
import Foundation
import SwiftUI

/// 一条关心关系
struct CareRelation: Codable, Identifiable, Equatable {
    var id: String          // 对方的 deviceId / 绑定码
    var name: String        // 昵称
    var bindCode: String    // 关心码（6位）
    var bindDate: Date      // 绑定日期
    var lastActive: Date?   // 最后活跃时间（从服务端同步）
    var isCharging: Bool = false

    /// 关心天数（从绑定日算起，最少 1 天）
    var days: Int {
        let d = Calendar.current.dateComponents([.day], from: bindDate, to: Date()).day ?? 0
        return max(d, 0) + 1
    }

    /// 活动状态文案
    var activityText: String {
        guard let ts = lastActive else { return "暂无活动" }
        let secondsAgo = Int(Date().timeIntervalSince(ts))
        switch secondsAgo {
        case ..<60:   return "刚刚活跃"
        case ..<3600: return "\(secondsAgo / 60) 分钟前活跃"
        case ..<86400: return "\(secondsAgo / 3600) 小时前活跃"
        default:      return "\(secondsAgo / 86400) 天前活跃"
        }
    }

    var isActive: Bool {
        guard let ts = lastActive else { return false }
        return Date().timeIntervalSince(ts) < 86400 // 24小时内算活跃
    }
}

/// 关心关系本地持久化管理
@MainActor
final class CareStore: ObservableObject {
    /// 我关心的人（关心列表）
    @Published var caring: [CareRelation] = []

    private let caringKey  = "anhao.care.caring"

    /// 被关心人数（由 MonitorManager 在心跳后更新）
    @Published var caredByCount: Int = 0

    /// 被关心天数（基于服务端返回人数计算）
    var totalCaredDays: Int {
        guard caredByCount > 0 else { return 0 }
        let earliest = caring.map(\.bindDate).min() ?? Date()
        let d = Calendar.current.dateComponents([.day], from: earliest, to: Date()).day ?? 0
        return max(d, 0) + 1
    }

    init() {
        load()
    }

    /// 从服务端恢复「我关心的人」（重装/换机后恢复；不覆盖已存在的本地关系）
    func syncFromServer() async {
        let devId = await Reporter.shared.deviceId
        let remote = await Reporter.shared.fetchCaring(deviceId: devId)
        guard !remote.isEmpty else { return }
        let existingCodes = Set(caring.map { $0.bindCode })
        let newOnes = remote.filter { !existingCodes.contains($0.bindCode) }
        guard !newOnes.isEmpty else { return }
        for r in newOnes {
            caring.append(CareRelation(
                id: UUID().uuidString,
                name: r.name.isEmpty ? r.bindCode : r.name,
                bindCode: r.bindCode,
                bindDate: Date()
            ))
        }
        save()
        print("[CareStore] synced \(newOnes.count) relations from server, total caring=\(caring.count)")
    }

    // MARK: - 我关心

    func addCaring(name: String, bindCode: String) {
        let rel = CareRelation(
            id: UUID().uuidString,
            name: name,
            bindCode: bindCode,
            bindDate: Date(),
            lastActive: nil
        )
        caring.append(rel)
        save()
        // 向服务端登记关心关系（不传昵称，隐私数据保留本地）
        Task {
            let ok = await Reporter.shared.registerCare(toCode: bindCode)
            print("[CareStore] registerCare result: \(ok)")
            // 添加后立即拉取状态
            await refreshCaredStatus()
        }
    }

    func removeCaring(_ relation: CareRelation) {
        caring.removeAll { $0.id == relation.id }
        save()
        // 同步删除服务端关心关系
        Task {
            let _ = await Reporter.shared.unregisterCare(toCode: relation.bindCode)
        }
    }

    func updateName(_ relation: CareRelation, name: String) {
        guard let idx = caring.firstIndex(where: { $0.id == relation.id }) else { return }
        caring[idx].name = name
        save()
    }

    // MARK: - 活动状态

    /// 拉取被关心者的最近活动状态
    func refreshCaredStatus() async {
        let codes = caring.map { $0.bindCode }
        guard !codes.isEmpty else { return }
        let statuses = await Reporter.shared.fetchCaredStatus(codes: codes)
        for i in caring.indices {
            let code = caring[i].bindCode
            if let s = statuses[code] {
                caring[i].lastActive = s.lastActive.map { Date(timeIntervalSince1970: TimeInterval($0)) }
                caring[i].isCharging = s.isCharging
            }
        }
    }

    // MARK: - 聚合

    var caringCount: Int { caring.count }

    var totalCaringDays: Int {
        guard let first = caring.map(\.bindDate).min() else { return 0 }
        let d = Calendar.current.dateComponents([.day], from: first, to: Date()).day ?? 0
        return max(d, 0) + 1
    }

    // MARK: - 持久化

    private func save() {
        guard let data = try? JSONEncoder().encode(caring),
              let json = String(data: data, encoding: .utf8) else { return }
        UserDefaults.standard.set(json, forKey: caringKey)
    }

    private func load() {
        guard let json = UserDefaults.standard.string(forKey: caringKey),
              let data = json.data(using: .utf8),
              let list = try? JSONDecoder().decode([CareRelation].self, from: data) else { return }
        caring = list
    }
}
